helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -147,10 +147,16 @@ protected static RealmAwareZkClient createZkClient(HelixManager manager) {
     if (Boolean.getBoolean(SystemPropertyKeys.MULTI_ZK_ENABLED) || zkAddress == null) {
       String clusterName = manager.getClusterName();
       String shardingKey = HelixUtil.clusterNameToShardingKey(clusterName);
+      RealmAwareZkClient.RealmAwareZkConnectionConfig zkConnectionConfig =
+          manager.getRealmAwareZkConnectionConfig();
       RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
           new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder()
               .setRealmMode(RealmAwareZkClient.RealmMode.SINGLE_REALM)
-              .setZkRealmShardingKey(shardingKey).build();
+              .setZkRealmShardingKey(shardingKey).setRoutingDataSourceEndpoint(
+              (zkConnectionConfig != null) ? zkConnectionConfig.getRoutingDataSourceEndpoint()
+                  : null).setRoutingDataSourceType(
+              (zkConnectionConfig != null) ? zkConnectionConfig.getRoutingDataSourceType() : null)
+              .build();
```

Comment:
How do we reason `setZkRealmShardingKey(shardingKey).setRoutingDataSourceEndpoint(...).setRoutingDataSourceType(...)`?

Note, this pattern is different from #1183 `configAccessor`.

Reply:
Please see PR description for why this is done. We need the routing data source endpoint from manager's connection config, because there're use cases where it's not provided among system properties.

Reply:
They why #1183 configAccessor does not have `setZkRealmShardingKey(shardingKey).setRoutingDataSourceEndpoint(...).setRoutingDataSourceType(...)`?

Reply:
Some of them (like ConfigAccessor) seem to be deprecated and for testing only, but some of them are not. I think this is a good question for @narendly : is it necessary to guard the other places with similar patterns, such as `ZkBucketDataAccessor`?

Reply:
I think the question is that what is the possible failure if we don't guard. But if you are considering if people will use it in the "wrong" way, then my comment is that when there is a way, someone will use it. And the bug will eventually be returned to us.
So as long as there is a possible hard failure case, please guard it. If it turns to be too complicated, then let's reconsider the overall design instead of patching here and there.

Reply:
Resolved offline with @jiajunwang: the design concern over ZooScalability is unrelated to this particular bug fix.